Framework/DataHandling/src/SaveAscii2.cpp

```diff
@@ -247,57 +249,88 @@ void SaveAscii2::exec() {
       }
     }
   }
-  if (!idx.empty()) {
-    nSpectra = static_cast<int>(idx.size());
+
+  // if no interval or spectra list, take all of them
+  if (idx.empty()) {
+    for (int i = 0; i < nSpectra; i++) {
+      idx.insert(i);
+    }
   }
 
   if (m_nBins == 0 || nSpectra == 0) {
     throw std::runtime_error("Trying to save an empty workspace");
   }
-  std::ofstream file(filename.c_str(), (appendToFile ? std::ios::app : std::ios::out));
 
-  if (!file) {
-    g_log.error("Unable to create file: " + filename);
-    throw Exception::FileError("Unable to create file: ", filename);
-  }
-  // Set the number precision
-  if (prec != EMPTY_INT()) {
-    file.precision(prec);
-  }
-  if (scientific) {
-    file << std::scientific;
-  }
-  const std::vector<std::string> logList = getProperty("LogList");
-  if (!logList.empty()) {
-    writeFileHeader(logList, file);
-  }
-  if (writeHeader) {
-    file << comment << " X " << m_sep << " Y " << m_sep << " E";
-    if (m_writeDX) {
-      file << " " << m_sep << " DX";
-    }
-    file << '\n';
-  }
+  bool OneSpectrumPerFile = getProperty("OneSpectrumPerFile");
+
+  Progress progress(this, 0.0, 1.0, idx.size());
+
   // populate the meta data map
   if (!m_metaData.empty()) {
     populateAllMetaData();
   }
-  if (idx.empty()) {
-    Progress progress(this, 0.0, 1.0, nSpectra);
-    for (int i = 0; i < nSpectra; i++) {
-      writeSpectrum(i, file);
-      progress.report();
+
+  auto idxIt = idx.begin();
+  while (idxIt != idx.end()) {
+    std::string currentFilename = filename;
+    size_t extPosition;
+    if (OneSpectrumPerFile) {
+      for (const std::string &ext : ASCII_EXTS) {
```

Comment:
What is this loop for? Can this be taken out? Indeed, these loops look cluttered, is there a good reason to use while, can we not use for or for each?

Reply:
I moved this to a dedicated function. The while loop is needed as the end is decided during the processing (either one file with all spectra or several files). The for loop is here to find the extension, I added a check to properly handle filenames that do not contain any extension.